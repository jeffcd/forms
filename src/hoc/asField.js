import React from 'react'
import get from 'lodash/get'
import FormContext from '../FormContext'
import VisibilityContext from '../VisibilityContext'
import ListContext from '../ListContext'
import validateField, { validatorFunctions } from '../validateField'
import messages from '../messages'
const validators = validatorFunctions

const getValidators = props => {
  const fieldValidators = {}
  Object.keys(props).forEach(name => {
    if (validators[name]) {
      fieldValidators[name] = props[name]
    }
  })
  return fieldValidators
}

const errorMessages = messages

const getErrorMessages = ({ props, field }) => {
  return field.errors.map(error => {
    let message = errorMessages[error]
    if (props.messages && props.messages[error]) {
      message = props.messages[error]
    }
    return message
      .replace('%name%', props.label)
      .replace('%param%', props[error])
  })
}

const getInitialValue = (
  fullName,
  name,
  Field,
  { minLength = 0 },
  data,
  initialValue
) => {
  const lookupPath = fullName.replace(/\.value/g, '')
  if (Field.type === 'list') {
    const dataList = get(data, lookupPath, [])
    const size = Math.max(dataList.length, parseInt(minLength))
    return Array(size)
  }
  return get(data, lookupPath, initialValue || '')
}

const asFieldHoc = Field => {
  return function asFieldHoc({ ...rest }) {
    return (
      <ListContext.Consumer>
        {listInfo => {
          const { name, initialValue, convertTo } = rest
          const fullName = listInfo
            ? `${listInfo.name}.value[${listInfo.i}].${name}`
            : name

          const listProps = {}
          if (listInfo) {
            const { i, name } = listInfo
            const fullName = `${name}.value[${i}].${rest.name}`
            const id = `${name}:${i}::${rest.name}:`
            listProps.id = id
            if (rest.name) {
              listProps.name = fullName
            }
          }

          return (
            <FormContext.Consumer>
              {form => {
                const field = get(form.fields, fullName)
                if (!field) {
                  const field = {
                    name,
                    value: getInitialValue(
                      fullName,
                      name,
                      Field,
                      rest,
                      form.data,
                      initialValue
                    ),
                    validators: getValidators(rest),
                    convertTo,
                    errors: []
                  }
                  const isUser = false
                  form.setFormField(field, fullName, isUser)

                  return null
                }

                return (
                  <VisibilityContext.Consumer>
                    {visibility => {
                      if (field.isVisible !== visibility.isVisible) {
                        const isUser = false
                        field.isVisible = visibility.isVisible
                        form.setFormField({ ...field }, fullName, isUser)
                      }
                      if (visibility.isVisible === false) {
                        return null
                      }

                      const handleChange = e => {
                        const field = get(form.fields, fullName)
                        const newField = {
                          ...field,
                          value: e.target.value
                        }
                        const errors = validateField(newField)
                        newField.errors = errors
                        form.setFormField(newField, fullName)
                      }
                      return (
                        <Field
                          onChange={handleChange}
                          value={field.value}
                          errors={getErrorMessages({ props: rest, field })}
                          {...rest}
                          {...listProps}
                        />
                      )
                    }}
                  </VisibilityContext.Consumer>
                )
              }}
            </FormContext.Consumer>
          )
        }}
      </ListContext.Consumer>
    )
  }
}

export default asFieldHoc
