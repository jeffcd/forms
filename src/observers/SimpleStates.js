import React from 'react'
import Typography from '@material-ui/core/Typography'
import { makeStyles } from '@material-ui/core/styles'
import withFormStateHoc from '../withFormStateHoc'

const stateStrs = {
  submit: 'Saving...',
  error: 'There are errors...',
  dirty: 'Unsaved changes...',
  pristine: ''
}

const useStyles = makeStyles(theme => ({
  SimpleStates: {
    marginTop: '16px',
    paddingLeft: theme.spacing(1)
  }
}))

const SimpleStates = ({ form }) => {
  const classes = useStyles()
  return (
    <Typography variant="body1" className={classes.SimpleStates}>
      {stateStrs[form.state]}
    </Typography>
  )
}

export default withFormStateHoc(SimpleStates)
