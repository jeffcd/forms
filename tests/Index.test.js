import React from 'react'
import Form from '../src'
import asField from '../src/hoc/asField'
import SimpleStates from '../src/observers/SimpleStates'

const Input = asField(
  ({
    onChange,
    onBlur = () => {},
    name,
    id,
    value,
    errors,
    label,
    helperText,
    required = false
  }) => {
    return (
      <div>
        <label htmlFor={name}>{label}</label>
        <br />
        <input
          id={`input-${id}`}
          name={name}
          value={value}
          onChange={e => onChange(e)}
          onBlur={e => onBlur(e)}
          label={label}
          required={required}
        />
        <span>{errors.length ? errors[0] : helperText}</span>
      </div>
    )
  }
)

it('Form: Pristine', () => {
  const submitSpy = jest.fn()
  let bodySpy = {}
  const form = (
    <Form
      onSubmit={(e, body, actions) => {
        bodySpy = body
        submitSpy()
        actions.pristine()
      }}
    >
      <Input id="firstName" name="firstName" required />
      <SimpleStates />
      <button type="submit" id="submit-button">
        Submit
      </button>
    </Form>
  )
  const tree = mount(form)

  expect(tree.html()).toMatchSnapshot()

  tree.find('form').simulate('submit')
  tree.find('#submit-button').simulate('click')
  expect(tree.html()).toMatchSnapshot()
  expect(submitSpy).toHaveBeenCalledTimes(0)

  tree
    .find('#input-firstName')
    .simulate('change', { target: { name: 'firstName', value: 'Jeff' } })
  tree.find('#submit-button').simulate('click')
  tree.find('form').simulate('submit')
  expect(bodySpy).toMatchSnapshot()
  expect(submitSpy).toHaveBeenCalledTimes(1)
})

it('Form: Error', () => {
  const submitSpy = jest.fn()
  let bodySpy = {}
  const form = (
    <Form
      onSubmit={(e, body, actions) => {
        bodySpy = body
        submitSpy()
        actions.error()
      }}
    >
      <Input id="firstName" name="firstName" required />
      <SimpleStates />
      <button type="submit" id="submit-button">
        Submit
      </button>
    </Form>
  )
  const tree = mount(form)

  expect(tree.html()).toMatchSnapshot()

  tree.find('form').simulate('submit')
  tree.find('#submit-button').simulate('click')
  expect(tree.html()).toMatchSnapshot()
  expect(submitSpy).toHaveBeenCalledTimes(0)

  tree
    .find('#input-firstName')
    .simulate('change', { target: { name: 'firstName', value: 'Jeff' } })
  tree.find('#submit-button').simulate('click')
  tree.find('form').simulate('submit')
  expect(bodySpy).toMatchSnapshot()
  expect(submitSpy).toHaveBeenCalledTimes(1)
})

it('Form: Error: Custom', () => {
  const submitSpy = jest.fn()
  let bodySpy = {}
  const form = (
    <Form
      onSubmit={(e, body, actions) => {
        bodySpy = body
        submitSpy()
        actions.error('server')
      }}
      stateStrs={{
        error_server: 'There was a server error. Please try again later.'
      }}
    >
      <Input
        id="firstName"
        name="firstName"
        required
        match="^Jeff"
        messages={{ match: 'Your name must begin with Jeff.' }}
      />
      <SimpleStates />
      <button type="submit" id="submit-button">
        Submit
      </button>
    </Form>
  )
  const tree = mount(form)

  expect(tree.html()).toMatchSnapshot()

  tree.find('form').simulate('submit')
  tree.find('#submit-button').simulate('click')
  expect(tree.html()).toMatchSnapshot()
  expect(submitSpy).toHaveBeenCalledTimes(0)

  tree
    .find('#input-firstName')
    .simulate('change', { target: { name: 'firstName', value: 'J' } })
  tree.find('#submit-button').simulate('click')
  tree.find('form').simulate('submit')
  expect(bodySpy).toMatchSnapshot()
  expect(submitSpy).toHaveBeenCalledTimes(0)

  tree
    .find('#input-firstName')
    .simulate('change', { target: { name: 'firstName', value: 'Jeff' } })
  tree.find('#submit-button').simulate('click')
  tree.find('form').simulate('submit')
  expect(bodySpy).toMatchSnapshot()
  expect(tree.html()).toMatchSnapshot()
  expect(submitSpy).toHaveBeenCalledTimes(1)
})

it('Form: Clear', () => {
  const submitSpy = jest.fn()
  let bodySpy = {}
  const form = (
    <Form
      onSubmit={(e, body, actions) => {
        bodySpy = body
        submitSpy()
        actions.clear()
      }}
    >
      <Input id="firstName" name="firstName" required />
      <SimpleStates />
      <button type="submit" id="submit-button">
        Submit
      </button>
    </Form>
  )
  const tree = mount(form)

  expect(tree.html()).toMatchSnapshot()

  tree.find('form').simulate('submit')
  tree.find('#submit-button').simulate('click')
  expect(tree.html()).toMatchSnapshot()
  expect(submitSpy).toHaveBeenCalledTimes(0)

  tree
    .find('#input-firstName')
    .simulate('change', { target: { name: 'firstName', value: 'Jeff' } })
  tree.find('#submit-button').simulate('click')
  tree.find('form').simulate('submit')
  expect(bodySpy).toMatchSnapshot()
  expect(submitSpy).toHaveBeenCalledTimes(1)
})

it('Form: Scope', () => {
  const submitSpy = jest.fn()
  let bodySpy = {}
  const form = (
    <Form
      onSubmit={(e, body, actions) => {
        bodySpy = body
        submitSpy()
        actions.clear()
      }}
    >
      <Input
        name="references[0].name"
        label="Your Reference Name #1"
        required
      />
      <Input name="references[0].phone" label="Phone" required />
      <Input name="references[1].name" label="Your Reference Name #2" />
      <Input name="references[1].phone" label="Phone" />

      <Input name="skills[0].first" label="Your Best Skill" required />
      <Input name="skills[1].second" label="Your Second Skill" />
      <SimpleStates />
      <button type="submit" id="submit-button">
        Submit
      </button>
    </Form>
  )
  const tree = mount(form)

  expect(tree.html()).toMatchSnapshot()

  tree.find('form').simulate('submit')
  tree.find('#submit-button').simulate('click')
  expect(tree.html()).toMatchSnapshot()
  expect(submitSpy).toHaveBeenCalledTimes(0)

  tree
    .find('input[name="references[0].name"]')
    .simulate('change', { target: { name: '', value: 'Jeff Decker' } })
  tree
    .find('input[name="references[0].phone"]')
    .simulate('change', { target: { name: '', value: '1234567890' } })
  tree
    .find('input[name="skills[0].first"]')
    .simulate('change', { target: { name: '', value: 'coding' } })
  tree.find('#submit-button').simulate('click')
  tree.find('form').simulate('submit')
  expect(tree.html()).toMatchSnapshot()
  expect(bodySpy).toMatchSnapshot()
  expect(submitSpy).toHaveBeenCalledTimes(1)
})
