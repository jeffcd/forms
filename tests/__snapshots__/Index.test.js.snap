// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form: Clear 1`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><label for=\\"firstName\\"></label><br><input id=\\"input-firstName\\" name=\\"firstName\\" required=\\"\\" value=\\"\\"><span></span></div><button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;

exports[`Form: Clear 2`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><label for=\\"firstName\\"></label><br><input id=\\"input-firstName\\" name=\\"firstName\\" required=\\"\\" value=\\"\\"><span>undefined is required.</span></div>There are errors...<button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;

exports[`Form: Clear 3`] = `
Object {
  "firstName": "Jeff",
}
`;

exports[`Form: Error 1`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><label for=\\"firstName\\"></label><br><input id=\\"input-firstName\\" name=\\"firstName\\" required=\\"\\" value=\\"\\"><span></span></div><button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;

exports[`Form: Error 2`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><label for=\\"firstName\\"></label><br><input id=\\"input-firstName\\" name=\\"firstName\\" required=\\"\\" value=\\"\\"><span>undefined is required.</span></div>There are errors...<button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;

exports[`Form: Error 3`] = `
Object {
  "firstName": "Jeff",
}
`;

exports[`Form: Error: Custom 1`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><label for=\\"firstName\\"></label><br><input id=\\"input-firstName\\" name=\\"firstName\\" required=\\"\\" value=\\"\\"><span></span></div><button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;

exports[`Form: Error: Custom 2`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><label for=\\"firstName\\"></label><br><input id=\\"input-firstName\\" name=\\"firstName\\" required=\\"\\" value=\\"\\"><span>undefined is required.</span></div>There are errors...<button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;

exports[`Form: Error: Custom 3`] = `Object {}`;

exports[`Form: Error: Custom 4`] = `
Object {
  "firstName": "Jeff",
}
`;

exports[`Form: Error: Custom 5`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><label for=\\"firstName\\"></label><br><input id=\\"input-firstName\\" name=\\"firstName\\" required=\\"\\" value=\\"Jeff\\"><span></span></div>There was a server error. Please try again later.<button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;

exports[`Form: Hydrate 1`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><label for=\\"references[0].name\\">Your Reference Name #1</label><br><input id=\\"input-undefined\\" name=\\"references[0].name\\" label=\\"Your Reference Name #1\\" required=\\"\\" value=\\"\\"><span></span></div><div><label for=\\"references[0].phone\\">Phone</label><br><input id=\\"input-undefined\\" name=\\"references[0].phone\\" label=\\"Phone\\" required=\\"\\" value=\\"\\"><span></span></div><div><label for=\\"references[1].name\\">Your Reference Name #2</label><br><input id=\\"input-undefined\\" name=\\"references[1].name\\" label=\\"Your Reference Name #2\\" value=\\"\\"><span></span></div><div><label for=\\"references[1].phone\\">Phone</label><br><input id=\\"input-undefined\\" name=\\"references[1].phone\\" label=\\"Phone\\" value=\\"\\"><span></span></div><div><label for=\\"skills[0].first\\">Your Best Skill</label><br><input id=\\"input-undefined\\" name=\\"skills[0].first\\" label=\\"Your Best Skill\\" required=\\"\\" value=\\"\\"><span></span></div><div><label for=\\"skills[1].second\\">Your Second Skill</label><br><input id=\\"input-undefined\\" name=\\"skills[1].second\\" label=\\"Your Second Skill\\" value=\\"\\"><span></span></div><button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;

exports[`Form: Hydrate 2`] = `
Object {
  "references": Array [
    Object {
      "name": "Jeff",
      "phone": "123456789",
    },
    Object {
      "name": "",
      "phone": "",
    },
  ],
  "skills": Array [
    Object {
      "first": "1st",
    },
    Object {
      "second": "2nd",
    },
  ],
}
`;

exports[`Form: Pristine 1`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><label for=\\"firstName\\"></label><br><input id=\\"input-firstName\\" name=\\"firstName\\" required=\\"\\" value=\\"\\"><span></span></div><button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;

exports[`Form: Pristine 2`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><label for=\\"firstName\\"></label><br><input id=\\"input-firstName\\" name=\\"firstName\\" required=\\"\\" value=\\"\\"><span>undefined is required.</span></div>There are errors...<button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;

exports[`Form: Pristine 3`] = `
Object {
  "firstName": "Jeff",
}
`;

exports[`Form: Scope 1`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><label for=\\"references[0].name\\">Your Reference Name #1</label><br><input id=\\"input-undefined\\" name=\\"references[0].name\\" label=\\"Your Reference Name #1\\" required=\\"\\" value=\\"\\"><span></span></div><div><label for=\\"references[0].phone\\">Phone</label><br><input id=\\"input-undefined\\" name=\\"references[0].phone\\" label=\\"Phone\\" required=\\"\\" value=\\"\\"><span></span></div><div><label for=\\"references[1].name\\">Your Reference Name #2</label><br><input id=\\"input-undefined\\" name=\\"references[1].name\\" label=\\"Your Reference Name #2\\" value=\\"\\"><span></span></div><div><label for=\\"references[1].phone\\">Phone</label><br><input id=\\"input-undefined\\" name=\\"references[1].phone\\" label=\\"Phone\\" value=\\"\\"><span></span></div><div><label for=\\"skills[0].first\\">Your Best Skill</label><br><input id=\\"input-undefined\\" name=\\"skills[0].first\\" label=\\"Your Best Skill\\" required=\\"\\" value=\\"\\"><span></span></div><div><label for=\\"skills[1].second\\">Your Second Skill</label><br><input id=\\"input-undefined\\" name=\\"skills[1].second\\" label=\\"Your Second Skill\\" value=\\"\\"><span></span></div><button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;

exports[`Form: Scope 2`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><label for=\\"references[0].name\\">Your Reference Name #1</label><br><input id=\\"input-undefined\\" name=\\"references[0].name\\" label=\\"Your Reference Name #1\\" required=\\"\\" value=\\"\\"><span>Your Reference Name #1 is required.</span></div><div><label for=\\"references[0].phone\\">Phone</label><br><input id=\\"input-undefined\\" name=\\"references[0].phone\\" label=\\"Phone\\" required=\\"\\" value=\\"\\"><span>Phone is required.</span></div><div><label for=\\"references[1].name\\">Your Reference Name #2</label><br><input id=\\"input-undefined\\" name=\\"references[1].name\\" label=\\"Your Reference Name #2\\" value=\\"\\"><span></span></div><div><label for=\\"references[1].phone\\">Phone</label><br><input id=\\"input-undefined\\" name=\\"references[1].phone\\" label=\\"Phone\\" value=\\"\\"><span></span></div><div><label for=\\"skills[0].first\\">Your Best Skill</label><br><input id=\\"input-undefined\\" name=\\"skills[0].first\\" label=\\"Your Best Skill\\" required=\\"\\" value=\\"\\"><span>Your Best Skill is required.</span></div><div><label for=\\"skills[1].second\\">Your Second Skill</label><br><input id=\\"input-undefined\\" name=\\"skills[1].second\\" label=\\"Your Second Skill\\" value=\\"\\"><span></span></div>There are errors...<button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;

exports[`Form: Scope 3`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><label for=\\"references[0].name\\">Your Reference Name #1</label><br><input id=\\"input-undefined\\" name=\\"references[0].name\\" label=\\"Your Reference Name #1\\" required=\\"\\" value=\\"\\"><span></span></div><div><label for=\\"references[0].phone\\">Phone</label><br><input id=\\"input-undefined\\" name=\\"references[0].phone\\" label=\\"Phone\\" required=\\"\\" value=\\"\\"><span></span></div><div><label for=\\"references[1].name\\">Your Reference Name #2</label><br><input id=\\"input-undefined\\" name=\\"references[1].name\\" label=\\"Your Reference Name #2\\" value=\\"\\"><span></span></div><div><label for=\\"references[1].phone\\">Phone</label><br><input id=\\"input-undefined\\" name=\\"references[1].phone\\" label=\\"Phone\\" value=\\"\\"><span></span></div><div><label for=\\"skills[0].first\\">Your Best Skill</label><br><input id=\\"input-undefined\\" name=\\"skills[0].first\\" label=\\"Your Best Skill\\" required=\\"\\" value=\\"\\"><span></span></div><div><label for=\\"skills[1].second\\">Your Second Skill</label><br><input id=\\"input-undefined\\" name=\\"skills[1].second\\" label=\\"Your Second Skill\\" value=\\"\\"><span></span></div><button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;

exports[`Form: Scope 4`] = `
Object {
  "references": Array [
    Object {
      "name": "Jeff Decker",
      "phone": "1234567890",
    },
    Object {
      "name": "",
      "phone": "",
    },
  ],
  "skills": Array [
    Object {
      "first": "coding",
    },
    Object {
      "second": "",
    },
  ],
}
`;
